Sources/SwiftParser/StringLiterals.swift

```diff
@@ -44,6 +148,243 @@ extension Parser {
+        let indentationByteLength = indentationTrivia.reduce(0, { $0 + $1.byteLength })
+        indentation = SyntaxText(rebasing: lastSegment.content.tokenText[0..<indentationByteLength])
+        middleSegments.append(.stringSegment(lastSegment))
+      } else {
+        indentationTrivia = []
+        indentation = ""
+      }
+
+      unexpectedBeforeCloseQuote = [closeQuote]
+      closeQuote = RawTokenSyntax(missing: closeQuote.tokenKind, leadingTriviaPieces: [.newlines(1)] + indentationTrivia, arena: self.arena)
+    }
+
+    // -------------------------------------------------------------------------
+    // Check open quote followed by newline
+
+    // Condition for the loop below that indicates whether the segment we are
+    // iterating over is on a new line.
+    var isSegmentOnNewLine: Bool
+
+    if let newlineTrivia = firstSegment?.content.tokenText.triviaPieceIfNewline {
+      openQuote = openQuote.extendingTrailingTrivia(by: [newlineTrivia], arena: self.arena)
+      isSegmentOnNewLine = true
+    } else {
+      if let firstSegment = firstSegment {
+        middleSegments.insert(.stringSegment(firstSegment), at: 0)
+      }
+      isSegmentOnNewLine = false
+      unexpectedBeforeOpenQuote = [openQuote]
+      openQuote = RawTokenSyntax(missing: openQuote.tokenKind, trailingTriviaPieces: [.newlines(1)] + indentationTrivia, arena: self.arena)
+    }
+
+    // -------------------------------------------------------------------------
+    // Check indentation of segments and escaped newlines at end of segment
+
+    let expressionIndentationChecker = StringLiteralExpressionIndentationChecker(expectedIndentation: indentation, arena: self.arena)
+
+    for (index, segment) in middleSegments.enumerated() {
+      switch segment {
+      case .stringSegment(var segment):
+        // We are not considering leading trivia for indentation computation.
+        // If these assertions are violated, we can probably lift them but we
+        // would need to check the produce the expected results.
+        assert(segment.content.leadingTriviaByteLength == 0)
+
+        // Re-classify indentation as leading trivia
+        if isSegmentOnNewLine {
+          if segment.content.tokenText.hasPrefix(indentation) {
+            segment = RawStringSegmentSyntax(
+              segment.unexpectedBeforeContent,
+              content: segment.content.reclassifyAsLeadingTrivia(indentationTrivia, arena: self.arena),
+              segment.unexpectedAfterContent,
+              arena: self.arena
+            )
+          } else {
+            let actualIndentation = segment.content.tokenText.prefix(while: { $0 == UInt8(ascii: " ") || $0 == UInt8(ascii: "\t") })
+            let actualIndentationTriva = TriviaParser.parseTrivia(SyntaxText(rebasing: actualIndentation), position: .leading)
+            let content = segment.content.reclassifyAsLeadingTrivia(
```

Comment:
Could this just take the length to reclassify and do the trivia parsing itself?

Reply:
No, we can't because `TriviaParser` lives in the SyntaxParser module whereas these functions are defined in the SwiftSyntax module. And if we wanted to pull them up into SyntaxParser, we would need to make scary implementation details (`rawData`, `RawTriviaPieceBuffer`, designated factory methods to create raw nodes) public.

Reply:
Thought about this more and spoke to Rintaro. Few questions:
1. Could we move this into `RawSyntaxTokenView` since these only apply to tokens?
2. To avoid a whole bunch of code, how would you feel if we always gave back parsed tokens instead? I'd even be fine with just asserting on the materialized case. But this way we could completely avoid having to parse the trivia + just pass in the length as above.